File: jlc_clojure/src/fundingcircle/jukebox/step_registry.clj
(ns fundingcircle.jukebox.step-registry
  "Step registry."
  (:refer-clojure :exclude [merge])
  (:import (java.util UUID)))

(defn create
  "Creates a step registry."
  []
  {:snippets {} :definitions [] :callbacks {} :resources []})

(defn definitions
  "Gets the step definitions that have been loaded."
  [step-registry]
  (:definitions step-registry))

(defn callbacks
  "Gets the step callbacks that have been found."
  [step-registry]
  (:callbacks step-registry))

(defn merge
  "Merge step registry."
  [step-registry language snippet definitions callbacks resources]
  (-> step-registry
      (update :snippets assoc language snippet)
      (update :definitions into definitions)
      (update :callbacks clojure.core/merge callbacks)
      (update :resources into resources)))

(defn register-step
  "Registers a step or hook definition."
  [step-registry {:keys [triggers opts callback]}]
  (let [id   (str (UUID/randomUUID))
        tags (:scene/tags opts)
        tags (if (string? tags) [tags] tags)
        opts (dissoc (assoc opts :scene/tags tags) :tags)]
    (-> step-registry

        (update :callbacks assoc id callback)
        (update :definitions conj {:id id :triggers triggers :opts opts}))))

(defn run
  "Run a step or hook."
  [step-registry {:keys [id board args] :as message}]
  (let [callback (get-in step-registry [:callbacks id])]
    (when-not callback (throw (ex-info "Undefined callback" {:message message})))
    (apply callback board args)))


(defn find-definition
  "Finds the step definition for the trigger."
  [step-registry trigger]
  (first
    (filter
      #(some (fn [t] (= trigger t)) (:triggers %))
      (:definitions step-registry))))

File: jlc_ruby/lib/jukebox/step_registry.rb
# frozen_string_literal: true

require 'active_support'
require 'active_support/core_ext'
require 'cucumber/cucumber_expressions/cucumber_expression'
require 'cucumber/cucumber_expressions/regular_expression'
require 'cucumber/cucumber_expressions/parameter_type_registry'
require 'logger'
require 'singleton'
require 'securerandom'
require 'set'

module Jukebox
  # Step registry.
  class StepRegistry
    include Singleton
    attr_reader :definitions, :callbacks, :snippets

    def initialize
      @snippets = {}
      @definitions = []
      @callbacks = {}
      @resources = Set[]
      @hooks = { before: [], after: [], before_step: [], after_step: [] }
      @board = { 'fundingcircle.jukebox.backend.cucumber' => true }
    end

    def valid_board?
      @board['fundingcircle.jukebox.backend.cucumber']
    end

    # Add a step or hook to the step registry.
    def add(*triggers, **opts, &block)
      raise UndefinedError unless block

      triggers.map! { |t| t.is_a?(Symbol) ? t : t.inspect[1..-2] }
      id = SecureRandom.uuid
      @definitions << {
        id: id,
        triggers: triggers,
        opts: StepRegistry.cleanup_opts(opts)
      }
      @callbacks[id] = block
      @resources += opts[:resources] if opts[:resources]
    end

    def register_callback(trigger, callback)
      if @hooks.key?(trigger)
        @hooks[trigger] << callback
      elsif trigger.is_a?(String)
        if trigger.starts_with?('^')
          expr = ::Cucumber::CucumberExpressions::RegularExpression.new(
            trigger,
            ::Cucumber::CucumberExpressions::ParameterTypeRegistry.new
          )
          @callbacks[expr] = callback
        else
          expr = ::Cucumber::CucumberExpressions::CucumberExpression.new(
            trigger,
            ::Cucumber::CucumberExpressions::ParameterTypeRegistry.new
          )
          @callbacks[expr] = callback
        end
      else
        raise "Don't know how to register callback with trigger: #{trigger}"
      end
    end

    # Runs a step
    def run(id:, board:, args:, **message)
      callback = @callbacks[id]
      raise UndefinedError, "Undefined callback: #{message}" unless callback

      # puts "Client running step: #{callback}: #{board} / #{args}"
      callback.call(board, *args)
    end

    def run_hooks(trigger)
      raise "Unknown hook type: #{trigger}" unless @hooks.key?(trigger)

      @hooks[trigger].each do |callback|
        scenario = {} # TODO: fill in scenario details
        board = callback.call(@board, scenario)
        raise "Dropped board: Before: '#{@board}' /  After: #{board}" unless valid_board?

        @board = board
      end
    end

    def run_step(trigger)
      callback, args = find_callback(trigger)
      raise "Don't know how to run step: #{trigger}" unless callback

      board = callback.call(@board, *args)
      raise "Dropped board: Before: '#{@board}' /  After: #{board}" unless valid_board?

      @board = board
    rescue => e
      puts "run_step exception: #{e}"
      raise e
    end

    # Finds a step's definition
    def find_definition(trigger)
      @definitions.find { |d| d[:triggers].include? trigger }
    end

    # Find a callback
    def find_callback(step_text)
      step_text = step_text.to_s
      _, callback = @callbacks.find { |expr, _|
        # puts "Finding callback '#{step_text}' / '#{expr}' #{expr.match(step_text)}"
        expr.match(step_text)
      }
      callback
    end

    # Normalizes a step's options
    def self.cleanup_opts(opts)
      opts['scene/tags'.to_sym] = [opts[:tags]] if opts[:tags].is_a?(String)
      opts['scene/resources'.to_sym] = opts[:resources] if opts[:resources]
      opts.except!(:tags, :resources)
    end
  end
end
